### PR TITLE
enhance: auto layout hasSider

### LINKED ISSUE #44518: layout组件嵌套 刷新会导致页面闪屏
### Reproduction link

[https://github.com/dr-forget/react-antd-bug.git](https://github.com/dr-forget/react-antd-bug.git)

### Steps to reproduce

npm run dev 运行项目 然后刷新浏览器即可  即可复现闪屏

### What is expected?

恢复正常  不闪屏 样式不错乱

### What is actually happening?

样式错乱 刷新会造成闪屏的效果

| Environment | Info |
| --- | --- |
| antd | 5.8.5 |
| React | 18.2.0 |
| System | macos13.4.1 |
| Browser | chrome 116.0.5845.110 |

### PR DESCRIPTION
**Pull request by zombieJ**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [x] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

resolves #44518 #42576

### 💡 Background and solution

遍历子节点添加 `hasSider`，不是最优解，比如对 HOC 无效。不过这个应该也能解90% 的 SSR 问题了。

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |     Layout support auto `hasSider` check to avoid blink in SSR.      |
| 🇨🇳 Chinese |    Layout 支持自动检测 `hasSider` 以防止在 SSR 场景下的闪烁问题。      |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

---

### <samp>🤖 Generated by Copilot at 1cf0df7</samp>

The pull request refactors the Layout and Sider components to use React hooks and a separate context file. It also improves the test coverage and the import order for the components. The main files affected are `layout.tsx`, `Sider.tsx`, `context.ts`, `useHasSider.ts`, and `index.test.tsx`.

### <samp>🤖 Generated by Copilot at 1cf0df7</samp>

* Extract and export `LayoutContext` and its interface from `layout.tsx` to a new file `context.ts` for better modularity and reusability ([link](https://github.com/ant-design/ant-design/pull/45361/files?diff=unified&w=0#diff-8b81b4752361ca6e5361e8f2412939ead23e04f52a5190dd11fa2169a811c6faR1-R15), [link](https://github.com/ant-design/ant-design/pull/45361/files?diff=unified&w=0#diff-f22153e3f63b1ae5f7b6d84d7ecbd52b795c4ccd564a452021da1205d7c17658L19-L31), [link](https://github.com/ant-design/ant-design/pull/45361/files?diff=unified&w=0#diff-c58876b8810c9f329249ebe0da66a0b3bc0953591d3fec9125ee7de20e9cf2cdL6-R11)).

### CODE REVIEW

**Review by MadCcc — APPROVED** [review]

File: components/layout/hooks/useHasSider.ts [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+    return hasSider;
+  }
+
+  if (siders.length) {
+    return true;
+  }
+
+  const childNodes = toArray(children);
+
+  return childNodes.some((node) => node.type === Sider);
```
Comment (kiner-tang) [inline components/layout/hooks/useHasSider.ts:21]: type 不是应该是字符串吗？Slider 应该是一个组件吧，这个能相等吗？

**Review by kiner-tang — APPROVED** [review]

**Comment by kiner-tang** [comment]
要更新一下快照

File: components/layout/hooks/useHasSider.ts (hunk 1 above)
Comment (zombieJ) [reply components/layout/hooks/useHasSider.ts:21]: type 是 ReactComponent 类型，包含字符串和方法和特殊 obj

**Comment by zombieJ** [comment]
snapshot 也修复了：

截屏2023-10-16 14 28 49

**Pull request merged**